types/vote.go

```diff
@@ -428,15 +428,15 @@ func SignAndCheckVote(
 		return false, &ErrVoteExtensionInvalid{ExtSignature: v.ExtensionSignature}
 	}
 
-	isNil := vote.BlockID.IsZero()
-	extSignature := (len(v.ExtensionSignature) > 0)
-	if extSignature == (!isPrecommit || isNil) {
-		// Non-recoverable because the vote is malformed
-		return false, &ErrVoteExtensionInvalid{ExtSignature: v.ExtensionSignature}
-	}
-
 	vote.ExtensionSignature = nil
 	if extensionsEnabled {
+		isNil := vote.BlockID.IsZero()
```

Comment:
This fix is disabling various sanity checks we wouldn't like to disable:
* Make sure that a prevote does not contain an extension signature
* Make sure that `nil` precommit does not contain an extension signature

Therefore this fix needs to "undo" the XOR in the if into two conditions, and only include one in the `if extensionsEnabled`.

I will post a diff here to explain my point

Reply:
This is how I believe the fix should look like:
```diff
diff --git a/types/vote.go b/types/vote.go
--- a/types/vote.go
+++ b/types/vote.go
@@ -430,13 +430,17 @@ func SignAndCheckVote(
 
 	isNil := vote.BlockID.IsZero()
 	extSignature := (len(v.ExtensionSignature) > 0)
-	if extSignature == (!isPrecommit || isNil) {
-		// Non-recoverable because the vote is malformed
+	if extSignature && (!isPrecommit || isNil) {
+		// Non-recoverable because the vote is malformed (wrongly contains an extension signature)
 		return false, &ErrVoteExtensionInvalid{ExtSignature: v.ExtensionSignature}
 	}
 
 	vote.ExtensionSignature = nil
 	if extensionsEnabled {
+		if !extSignature && isPrecommit && !isNil {
+			// Non-recoverable because the vote is malformed (extension signature missing for a precommit)
+			return false, &ErrVoteExtensionInvalid{ExtSignature: v.ExtensionSignature}
+		}
 		vote.ExtensionSignature = v.ExtensionSignature
 	}
 
```

Reply:
**However**, this fix may only be buying you time. If the chain decides to start using vote extensions (in the SDK, it would be via a gov proposal)... and your remote signer still doesn't sign vote extensions, it will start failing again.

Reply:
Agree.

Reply:
> Does the main branch also need to check this condition?

Thanks for pointing out. Taking care of it (#3642)